### PR TITLE
[FLINK-27876][table-planner] Choose the right side as build side when using default shuffle hash strategy if left size is equal with right

### PR DESCRIPTION
**Pull request by xuyangzhong**
## What is the purpose of the change

 If leftSize or rightSize is unknown or equal in Semi/Anti Join, the right side of the join should be chosen as build side.

## Brief change log

  - Choose the right side in `BatchPhysicalHashJoinRule` when using default shuffle hash strategy

## Verifying this change

Existent test cases can cover this change.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented?

### CODE REVIEW

**Comment by xuyangzhong** [comment]
@flinkbot run azure

**Comment by xuyangzhong** [comment]
@flinkbot run azure

**Comment by xuyangzhong** [comment]
@flinkbot run azure

**Comment by xuyangzhong** [comment]
@flinkbot run azure

**Comment by xuyangzhong** [comment]
@flinkbot run azure

**Review by lsyldliu — COMMENTED** [review]
@xuyangzhong Thanks for your contribution, the changes look good to me. I just left a minor comment. Moreover, here exists conflicts with master branch, please fix it first.

File: flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/batch/sql/MultipleInputCreationTest.xml [hunk 1]
```diff
@@ -109,39 +109,6 @@ MultipleInput(readOrder=[0,1,0], members=[\nNestedLoopJoin(joinType=[LeftOuterJo
 :  +- LegacyTableSourceScan(table=[[default_catalog, default_database, x, source: [TestTableSource(a, b, c, nx)]]], fields=[a, b, c, nx])
 +- Exchange(distribution=[broadcast])
    +- LegacyTableSourceScan(table=[[default_catalog, default_database, y, source: [TestTableSource(d, e, f, ny)]]], fields=[d, e, f, ny])
-]]>
-    </Resource>
-  </TestCase>
-  <TestCase name="testIncludeCalcForChainableSource[shuffleMode: ALL_EXCHANGES_BLOCKING, schedulerType: AdaptiveBatch]">
```
[the remark this answers is not in the document]
Comment (xuyangzhong) [reply flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/batch/sql/MultipleInputCreationTest.xml:115]: Actually this test is not deleted. I just delete this xml, let Flink to generate it again, and the order about tests in this file has been changed.

**Comment by xuyangzhong** [comment]
Hi, @lsyldliu. I rebased the master. Could you have a look again?

**Review by lsyldliu — APPROVED** [review]
LGTM

**Comment by lsyldliu** [comment]
@xuyangzhong Do you think we should pick this fix to release-1.18?

**Comment by xuyangzhong** [comment]
> @xuyangzhong Do you think we should pick this fix to release-1.18?

I feel that cp is not needed because it changes the plan and will cause compatibility issues. WDYT?

**Pull request merged**